### PR TITLE
[OTX] otx find & build with OMZ backbones + Refactor codes

### PR DESCRIPTION
**Pull request by harimkang**
## Summary
Implemented otx find & build for OMZ backbone and refactored existing find and build code.

### What this PR contains
- Find OMZ backbones
image
- Classification build with OMZ backbones
- Classification build with mmdet, mmseg, torchvision backbones
- Refactoring otx find & build codes

### CODE REVIEW

File: otx/cli/builder/supported_backbone/mmdet.json
Comment (sungmanc) [inline otx/cli/builder/supported_backbone/mmdet.json:17]: How about unifying all JSON files into single file? (i.e. `otx/cli/builder/supported_backbones.json`)
I'm not sure whether the location of JSON files that has a dependency on frameworks is good or not since they are located under the `otx/cli` and `builder` that could be represented such a general term.

File: otx/cli/builder/builder.py
Comment (sungmanc) [inline otx/cli/builder/builder.py]: How about converting all `string variables` to the `TaskType` class?
(i.e. "classification" --> "TaskType.CLASSIFICATION")

Then, we can represents all kinds of tasks supported by Geti and also can avoid hard-coded string format.

File: otx/cli/builder/builder.py
Comment (sungmanc) [inline otx/cli/builder/builder.py:284]: Maybe, we need to consider Action tasks, in the near future.
So, I just prefer the `elif` sentence in here to handle the `detection` and `segmentation`

File: otx/algorithms/common/adapters/mmcv/models/backbones/torchvision_backbones.py
Comment (sungmanc) [inline otx/algorithms/common/adapters/mmcv/models/backbones/torchvision_backbones.py:134]: Just wondering: can we handle all forward function implemented on torchvision backbones by using above 3 functions? (resnet, shufflenet, multioutput forward)

File: otx/cli/builder/builder.py
Comment (harimkang) [reply otx/cli/builder/builder.py]: Perhaps I could use something like `str(TaskType.CLASSIFICATION)` instead of `classification`. I'll fix it.

File: otx/cli/builder/supported_backbone/mmdet.json
Comment (harimkang) [reply otx/cli/builder/supported_backbone/mmdet.json:17]: There are parts that I think this is also a design choice. Agreeing with some of the comments, the downside to keeping it in one file is that it doesn't have good visibility. One file becomes very long and I suspect that if there is a problem with that file it could affect the whole framework's backbone find and build. what do you think?

File: otx/cli/builder/builder.py
Comment (harimkang) [reply otx/cli/builder/builder.py:284]: yes, I think this is something to consider. I will test it and if it is not a problem, I will fix it with `elif` & `else` (raise Error).

File: otx/algorithms/common/adapters/mmcv/models/backbones/torchvision_backbones.py
Comment (harimkang) [reply otx/algorithms/common/adapters/mmcv/models/backbones/torchvision_backbones.py:134]: Most `torchvision backbones` use layers in `self.layers` or `self.features`. In such cases, it is patched to use `multioutput forward`, and resnet and shufflenet are configured slightly differently, so additional forward patching was required.

**Review by jaegukhyun — DISMISSED** [review]
I have no big objection for this PR, it looks good to me. As otx build and otx find support all task, all training type, and even all training backends(mmX, torchvision, openvino), it will contain many if, else statements. Therefore, at some point, it will be needed to re-design builder.py for more scalable and simple codes.

**Comment by harimkang** [comment]
> I have no big objection for this PR, it looks good to me. As otx build and otx find support all task, all training type, and even all training backends(mmX, torchvision, openvino), it will contain many if, else statements. Therefore, at some point, it will be needed to re-design builder.py for more scalable and simple codes.

Yes, this is a PR that includes a re-design to reduce the `if` and `else` statements you mentioned in the current state. Nevertheless, There may still be some areas that are not general, but I think we can gradually improve this. thanks for the comment :)

File: otx/cli/builder/builder.py
Comment (harimkang) [reply otx/cli/builder/builder.py:284]: As a result of checking this, there are cases where there is no task variable for each model. Looks like I need to align this later. So I think we should leave it as it is for now.

File: otx/cli/builder/supported_backbone/mmdet.json
Comment (sungmanc) [reply otx/cli/builder/supported_backbone/mmdet.json:17]: Yeah, it can differ from person to person. Let's leave it as is.

**Review by JihwanEom — DISMISSED** [review]
LGTM

File: otx/cli/utils/importing.py
Comment (eunwoosh) [inline otx/cli/utils/importing.py:56]: How about deleting this line?

File: otx/cli/utils/importing.py
Comment (eunwoosh) [inline otx/cli/utils/importing.py]: How about combine this line with right above code as below.
available_backbone_path = os.path.join(get_otx_root_path(), "cli/builder/supported_backbone", f"{backend}.json")

File: otx/cli/utils/tests.py
Comment (eunwoosh) [inline otx/cli/utils/tests.py:626]: how about adding some asserts code to check whether code work as expected?

File: otx/cli/tools/find.py
Comment (eunwoosh) [inline otx/cli/tools/find.py]: How about adding type hinting?

File: otx/cli/utils/importing.py
Comment (harimkang) [reply otx/cli/utils/importing.py:56]: Your review has been reflected. thank you :)

File: otx/cli/utils/importing.py
Comment (harimkang) [reply otx/cli/utils/importing.py]: It's corrected.

File: otx/cli/utils/tests.py
Comment (harimkang) [reply otx/cli/utils/tests.py:626]: Added this.

File: otx/cli/tools/find.py
Comment (harimkang) [reply otx/cli/tools/find.py]: Added this too.

**Review by eunwoosh — APPROVED** [review]
LGTM. It can be merged after checking Pre-merge test

**Review by JihwanEom — APPROVED** [review]

**Comment by harimkang** [comment]
As a result of locally double-checking the failed cases in the CI test, it is expected that this PR is not a problem.

- Tiling Instance-Segmentation
image
- Segmentation
image

I will merge this.

**Pull request merged**
